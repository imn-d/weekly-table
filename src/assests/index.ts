export * from './CancelIcon';
export * from './DeleteIcon';
export * from './EditIcon';
export * from './DragIcon';
export * from './CheckIcon';
