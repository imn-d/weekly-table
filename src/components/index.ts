export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
