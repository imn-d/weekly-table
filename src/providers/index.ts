export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
