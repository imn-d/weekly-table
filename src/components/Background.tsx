import React, { memo, useEffect, useRef } from 'react';

import { useCells, useScheduler } from '../providers';

/**
 * Drawing cells background component
 */
export const Background = memo(() => {
  const { helperWidthProp, headerHeightProp, baseZIndex, bottomHeightProp } =
    useScheduler();
  const { widthStep, heightStep, offsetHeight, offsetWidth } = useCells();

  /**
   * Canvas reference
   */
  const canvasRef = useRef<HTMLCanvasElement>(null);

  /**
   * Context reference
   */
  const contextRef = useRef<CanvasRenderingContext2D | null>(null);

  /**
   * Cells redrawing
   */
  useEffect(() => {
    const canvas = canvasRef?.current;
    if (!canvas) return;
    canvas.width = offsetWidth * 2;
    canvas.height = offsetHeight * 2;
    canvas.style.width = `${offsetWidth}px`;
    canvas.style.height = `${offsetHeight}px`;
    const context = canvas.getContext('2d');
    if (!context) return;

    context.scale(2, 2);
    contextRef.current = context;
    contextRef.current.beginPath();
    for (let x = helperWidthProp; x <= offsetWidth; x += widthStep) {
      contextRef.current.moveTo(x, 0);
      contextRef.current.lineTo(x, offsetHeight - bottomHeightProp);
    }
    contextRef.current.strokeStyle = 'rgba(0,0,0, 0.2)';
    contextRef.current.lineWidth = 1;
    contextRef.current.stroke();
    contextRef.current.beginPath();
    for (
      let y = headerHeightProp;
      y <= offsetHeight - bottomHeightProp;
      y += heightStep
    ) {
      contextRef.current.moveTo(helperWidthProp, y);
      contextRef.current.lineTo(offsetWidth, y);
    }
    contextRef.current.strokeStyle = 'rgba(0,0,0, 0.1)';
    contextRef.current.lineWidth = 1;
    contextRef.current.stroke();
    contextRef.current.save();
  }, [
    bottomHeightProp,
    headerHeightProp,
    heightStep,
    helperWidthProp,
    offsetHeight,
    offsetWidth,
    widthStep,
  ]);

  return (
    <canvas
      id={'backgroundCanvas'}
      className={'scheduler-background-canvas'}
      style={{
        position: 'absolute',
        zIndex: baseZIndex + 1,
      }}
      ref={canvasRef}
    />
  );
});
