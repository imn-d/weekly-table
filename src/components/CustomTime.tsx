import React, { FC } from 'react';

import { TIME } from '../common';
import { useCells, useCustomTime, useScheduler } from '../providers';

/**
 * Custom time setter component
 */
export const CustomTime: FC = () => {
  const { baseZIndex } = useScheduler();
  const { widthStep } = useCells();
  const { isEditing, position, displayTime, handleTime, isError } =
    useCustomTime();

  return (
    <div
      id={'customTimeWrapper'}
      className={'scheduler-text custom-time-wrapper'}
      style={{
        display: isEditing ? 'flex' : 'none',
        zIndex: baseZIndex + 6,
        position: 'absolute',
        flexDirection: 'column',
        top: position?.top,
        left: position?.left,
        width: widthStep ? widthStep * 0.75 : 0,
        maxWidth: widthStep ? widthStep * 0.75 : 0,
      }}
    >
      {isError ? (
        <span
          id={'customTimeError'}
          className={'custom-time-error'}
          style={{
            position: 'absolute',
            top: 55,
            width: widthStep,
          }}
        >
          {isError}
        </span>
      ) : (
        <></>
      )}
      <input
        id={'customTimeInputStart'}
        className={'custom-time-input custom-time-input-start'}
        style={{
          width: widthStep ? widthStep * 0.75 : 0,
          maxWidth: widthStep ? widthStep * 0.75 : 0,
        }}
        type={'time'}
        value={displayTime.startTime}
        step={1}
        onChange={(event) =>
          handleTime(event?.target?.valueAsNumber, TIME.START)
        }
      />
      <input
        id={'customTimeInputEnd'}
        className={'custom-time-input custom-time-input-end'}
        style={{
          width: widthStep ? widthStep * 0.75 : 0,
          maxWidth: widthStep ? widthStep * 0.75 : 0,
        }}
        type={'time'}
        value={displayTime.endTime}
        step={1}
        onChange={(event) => handleTime(event?.target?.valueAsNumber, TIME.END)}
      />
    </div>
  );
};
