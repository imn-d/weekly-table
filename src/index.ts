import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
