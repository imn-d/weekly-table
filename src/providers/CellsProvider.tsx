import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { Cell, CellsProps, HOUR_24 } from '../common';
import { useScheduler } from './Scheduler';

const CellsContext = createContext<CellsProps>({} as CellsProps);
/**
 *  Virtual cells provider context hook
 *  @returns [CellsProps]{@link CellsProps}
 */
export const useCells = (): CellsProps => useContext(CellsContext);

/**
 * Virtual cells provider
 */
const CellsProvider: FC = ({ children }) => {
  const {
    parentRef,
    timeframe,
    columns,
    helperWidthProp,
    headerHeightProp,
    bottomHeightProp,
  } = useScheduler();

  /**
   * Main context state, setting by useEffect
   */
  const [cellsProps, setCellsProps] = useState<CellsProps>({} as CellsProps);

  /**
   * Main context updater
   */
  const setCells = useCallback(
    (update: Cell[]) => {
      setCellsProps({ ...cellsProps, cells: update });
    },
    [cellsProps],
  );

  /**
   * Ms count per timeframe
   */
  const msTime = useMemo(() => {
    return timeframe * 60 * 1000;
  }, [timeframe]);

  /**
   * Calculates context values and build virtual cells area
   */
  useEffect(() => {
    const parent = parentRef?.current;
    if (!parent) return;
    const widthStep = (parent.offsetWidth - helperWidthProp) / columns?.length;
    const areaHeight =
      parent.offsetHeight - headerHeightProp - bottomHeightProp;
    const heightStep = areaHeight / (1440 / timeframe);
    const millisPerPixel = HOUR_24 / areaHeight;
    const maxCellIndex = areaHeight / heightStep - 1;
    const cells = new Array<Cell>();
    let widthPos = helperWidthProp;
    for (let i = 0; i < columns?.length; i++) {
      let heightPos = headerHeightProp;
      for (let j = 0; j < areaHeight / heightStep; j++) {
        cells.push({
          isSelected: false,
          column: i,
          row: j,
          position: {
            x: widthPos,
            y: heightPos,
            w: widthPos + widthStep,
            h: heightPos + heightStep,
          },
        } as Cell);
        heightPos += heightStep;
      }
      widthPos += widthStep;
    }

    setCellsProps({
      widthStep,
      heightStep,
      millisPerPixel,
      cells,
      maxCellIndex,
      offsetWidth: parent.offsetWidth,
      offsetHeight: parent.offsetHeight,
    } as CellsProps);
  }, [
    parentRef?.current?.offsetWidth,
    parentRef?.current?.offsetHeight,
    columns?.length,
    bottomHeightProp,
    headerHeightProp,
    helperWidthProp,
    timeframe,
    parentRef,
  ]);

  return (
    <CellsContext.Provider
      value={{
        ...cellsProps,
        setCells: setCells,
        msTime,
      }}
    >
      {children}
    </CellsContext.Provider>
  );
};

export default CellsProvider;
