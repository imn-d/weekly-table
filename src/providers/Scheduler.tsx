import React, {
  FC,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react';

import {
  BlockColorsProps,
  SchedulerInputProps,
  SchedulerProps,
  schedulerColumns,
  schedulerRows,
} from '../common';
import { Background, CustomTime, Popup, SchedulerLayout } from '../components';
import CanvasProvider from './CanvasProvider';
import CellsProvider from './CellsProvider';
import CustomTimeProvider from './CustomTimeProvider';
import PointerLockProvider from './PointerLockProvider';
import PopupProvider from './PopupProvider';
import TimeBlockProvider from './TimeBlockProvider';

const SchedulerContext = createContext<SchedulerProps>({} as SchedulerProps);
/**
 *  Scheduler provider context hook
 *  @returns [SchedulerProps]{@link SchedulerProps}
 */
export const useScheduler = (): SchedulerProps => useContext(SchedulerContext);

/**
 * Root scheduler components
 * @param props{@link SchedulerInputProps} initial props
 */
const Scheduler: FC<SchedulerInputProps> = (props: SchedulerInputProps) => {
  const [update, setUpdate] = useState<number>(0);

  useEffect(() => {
    window.addEventListener('resize', () => setUpdate((update) => update + 1));
    return () => {
      window.removeEventListener('resize', () =>
        setUpdate((update) => update + 1),
      );
    };
  }, []);

  return (
    <SchedulerContext.Provider
      value={{
        parentRef: props.parentRef,
        blockColors: props?.blockColors
          ? props.blockColors
          : ({
              common: '#ff5722',
              temp: '#c6a700',
              draw: '#ff8a50',
              hover: '#ff3d00',
            } as BlockColorsProps),
        headerHeightProp: props?.headerHeightProp ? props.headerHeightProp : 80,
        helperWidthProp: props?.helperWidthProp ? props.helperWidthProp : 80,
        bottomHeightProp: props?.bottomHeightProp ? props.bottomHeightProp : 20,
        baseZIndex: props?.baseZIndex ? props.baseZIndex : 0,
        timeframe: props?.timeframe ? props.timeframe : 60,
        columns: props?.columns ? props.columns : schedulerColumns,
        rows: props?.rows ? props.rows : schedulerRows,
        requiredTZOffset: props?.requiredTZOffset,
        mouseSpeed: props?.mouseSpeed ? props.mouseSpeed : 3,
        defaultValue: props?.defaultValue,
        onChange: props?.onChange,
      }}
    >
      <CellsProvider>
        <CanvasProvider>
          <PointerLockProvider>
            <TimeBlockProvider>
              <PopupProvider>
                <CustomTimeProvider>
                  <Background />
                  <SchedulerLayout />
                  <Popup />
                  <CustomTime />
                </CustomTimeProvider>
              </PopupProvider>
            </TimeBlockProvider>
          </PointerLockProvider>
        </CanvasProvider>
      </CellsProvider>
    </SchedulerContext.Provider>
  );
};

export default Scheduler;
