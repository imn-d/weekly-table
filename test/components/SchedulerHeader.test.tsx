import { mount } from 'enzyme';
import React from 'react';

import {
  CellsProps,
  SchedulerColumnsProps,
  SchedulerProps,
} from '../../src/common';
import { SchedulerHeader } from '../../src/components';
import * as uc from '../../src/providers/CellsProvider';
import * as us from '../../src/providers/Scheduler';

describe('SchedulerHeader Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useCellMock = jest.spyOn(uc, 'useCells');

  beforeEach(() => {
    jest.clearAllMocks();

    useCellMock.mockReturnValue({
      widthStep: 20,
    } as CellsProps);
  });

  it('Base Render && Columns Count Test', () => {
    useSchedulerMock.mockReturnValueOnce({
      headerHeightProp: 80,
      columns: new Array<SchedulerColumnsProps>(
        {
          weight: 1,
          full: 'First',
          short: 'F',
        },
        {
          weight: 2,
          full: 'Second',
          short: 'S',
        },
      ),
    } as SchedulerProps);

    const wrapper = mount(<SchedulerHeader />);

    expect(useSchedulerMock).toBeCalledTimes(1);
    expect(useCellMock).toBeCalledTimes(1);

    expect(wrapper).toMatchSnapshot('baseRender');
    expect(wrapper.find('#schedulerHeaderColumn').hostNodes()).toHaveLength(2);
    wrapper.unmount();
  });

  it('Columns Undefined', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    useSchedulerMock.mockReturnValueOnce({
      headerHeightProp: 80,
      columns: undefined,
    } as SchedulerProps);

    const wrapper = mount(<SchedulerHeader />);
    expect(wrapper).toHaveLength(1);
    expect(wrapper.find('#schedulerHeaderColumn').hostNodes()).toHaveLength(0);
    wrapper.unmount();
  });
});
