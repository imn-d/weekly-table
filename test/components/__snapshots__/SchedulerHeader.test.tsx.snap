// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchedulerHeader Component Base Render && Columns Count Test: baseRender 1`] = `
<SchedulerHeader>
  <div
    className="scheduler-header"
    id="schedulerHeader"
    style={
      Object {
        "display": "inline-flex",
        "height": "80px",
        "maxHeight": "80px",
        "position": "relative",
        "userSelect": "none",
      }
    }
  >
    <div
      className="scheduler-text scheduler-header-column"
      id="schedulerHeaderColumn"
      key="1"
      style={
        Object {
          "display": "inline-flex",
          "minWidth": "20px",
          "width": "20px",
        }
      }
    >
      First
    </div>
    <div
      className="scheduler-text scheduler-header-column"
      id="schedulerHeaderColumn"
      key="2"
      style={
        Object {
          "display": "inline-flex",
          "minWidth": "20px",
          "width": "20px",
        }
      }
    >
      Second
    </div>
  </div>
</SchedulerHeader>
`;
